docs/developer/api-versioning.md

```diff
@@ -0,0 +1,388 @@
+
+Without loss of generality, we only consider additions and deletions
+of routes in this section: if you want to change the path or verb of
+an end-point, add a new path instead, and phase the old one out (now
+or in some future version).
+
+When end-points are present in some supported versions, but not in
+others, their record fields in the servant routing type needs to be
+present for all versions, but in some versions should behave as if it
+weren't.
+
+This is best solved by a new type alias:
+
+```haskell
+type NotInThisVersion = Verb 'NOTINTHISVERSION '[] NoContent
+```
+
+Then we can write a type family that can crawl a `ServantAPI` (not the
+record one, the one with `:<|>`) and drop all the routes marked as not
+existing.
+
+This will save us the trouble of writing lots of instances for
+`NotInThisVersion` (server, swagger, client, ...), and yield exactly
+the desired result:
+
+```haskell
+type ServantAPI =
+  DropNotInThisVersion
+    ( ToServantApi (Api "v1")
+        :<|> ToServantApi (Api "v2")
+        :<|> ToServantApi (Api "v4") -- v3 is broken
+    )
+```
+
+
+## Adoption of versioned APIs
+
+When API versions are introduced to a code base that has a routing
+table without versions, the question arises what to do with old
+clients or servers talking to new ones.
+
+We define a middleware that
+
+ (1) maps requests without version prefix in the path to ones that
+     have version `"v0"`.
+
+ (2) responds with a specific type of error if an unsupported version
+     is requested (so the client can re-negotiate a new version to
+     speak after an upgrade, see below).
+
+
+## Version handshake
+
+Client and server need to agree on a version to use.  The server
+provides two (kinds of) end-points for that.
+
+```
+GET /api-versions
+=> { "supported": [1, 2, 45, 119] }
+
+GET /v*/api-docs
+=> <swagger for this particular version>
+```
+
+The client developer can pull the swagger docs of a new version and
+diff it against the one they already support, and work their way
+through the changes (see below).
+
+The client will call `GET /api-versions` and pick any number in the
+intersection of the versions it supports and the ones the server
+supports (usually the largest).
+
+*Corner case:* if we want to distinguish between backend-to-backend
+and client-to-backend, we can do that in path suffixes (`GET
+/api-versions/{client,federation}` etc.).
+
+
+### No shared api version
+
+If the intersection set is empty, the client has no way of talking to
+the server.  It needs to politely fail with an error message about
+upgrading, downgrading, or talking to another server.
+
+This should only happen if the distance between last upgrade on client
+and server exceeds the agreed-upon limits (eg., 6 months).
+
+
+### Update detection and version re-negotation
+
+If the server is upgraded and some old supported versions are phased
+out, the client may be caught by surprise.
+
+The clients should therefore frequently (every 24h?) refetch
```

Comment:
clients can't perform time scheduled checks.

Reply:
ah, yes, this is outdated.  thanks, i'll fix that!

Reply:
0210409